handler/handler.go

```diff
@@ -0,0 +1,330 @@
+package handler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"html"
+	"log/slog"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/lsst-dm/s3nd/conf"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awshttp "github.com/aws/aws-sdk-go-v2/aws/transport/http"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	smithy "github.com/aws/smithy-go"
+	"github.com/google/uuid"
+	"github.com/hyperledger/fabric/common/semaphore"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+var (
+	logger                  = slog.New(slog.NewJSONHandler(os.Stdout, nil))
+	errUploadAttemptTimeout = errors.New("upload attempt timeout")
+)
+
+type S3ndHandler struct {
+	conf            *conf.S3ndConf
+	awsConfig       *aws.Config
+	s3Client        *s3.Client
+	uploader        *manager.Uploader
+	parallelUploads *semaphore.Semaphore
+}
+
+type uploadTask struct {
+	Id        uuid.UUID   `json:"id"`
+	Uri       *requestURL `json:"uri,omitempty"`
+	Bucket    *string     `json:"-"`
+	Key       *string     `json:"-"`
+	File      *string     `json:"file,omitempty"`
+	StartTime time.Time   `json:"-"`
+	EndTime   time.Time   `json:"-"`
+	Duration  string      `json:"duration,omitempty"`
+	Attempts  int         `json:"attempts,omitzero"`
+}
+
+func newUploadRequest(startTime time.Time) *uploadTask {
+	return &uploadTask{
+		Id:        uuid.New(),
+		StartTime: startTime,
+	}
+}
+
+func (t *uploadTask) stop() {
+	t.EndTime = time.Now()
+	t.Duration = t.EndTime.Sub(t.StartTime).String()
```

Comment:
Not sure why you want this Duration to be a string rather than a Duration you convert to a string when you need that, but I'm going to assume you have a reason.

Reply:
It is so the output has a unit on it: 
```
{"code":200,"msg":"upload success","task":{"id":"8f2250ba-29a9-4c6f-a8a1-930373a404fe","uri":"s3://bar/foo","file":"/tmp/foo","duration":"21.69503ms","attempts":1}}
```

Reply:
Otherwise, the value is an integer in nanoseconds.

Reply:
Oh, right, now I remember that about it from work I did on Telegraf.